File: Blind75-Practice/easy/12-1d-dp-l/arrays/Java/HouseRobberEasy.java
/*
LeetCode 70. Climbing Stairs
Difficulty: Easy
Pattern: Dynamic Programming / Fibonacci

You are climbing a staircase with n steps.
Each time, you can climb either 1 step or 2 steps.
Return the number of distinct ways to reach the top.

Examples:
n = 2 -> 2
Ways: [1+1], [2]

n = 3 -> 3
Ways: [1+1+1], [1+2], [2+1]

Idea:
The number of ways to reach step i is the sum of the ways to reach
step i - 1 and step i - 2.

Time: O(n), Space: O(1)
*/
public class HouseRobberEasy {
    public int climbStairs(int n) {
        int dp0 = 1;
        int dp1 = 1;
        for (int i = 2; i <= n; i++) {
            int curr = dp0 + dp1;
            dp0 = dp1;
            dp1 = curr;
        }

        return dp1;

        //  

        //

        //

        //

        //

        //

        //

        //

        //

        //

        //

        //

        //

        //

        //

        //

        //

        //

        //

        //

        //

        //

        //

        //

        //

        //

        //

        //

        //

        //

        // Fibonacci-style DP: ways[i] = ways[i-1] + ways[i-2]

        // if (n <= 2) {
        //     return n;
        // }
        // int prev2 = 1;
        // int prev1 = 2;
        // for (int i = 3; i <= n; i++) {
        //     int current = prev1 + prev2;
        //     prev2 = prev1;
        //     prev1 = current;
        // }
        // return prev1;
        throw new UnsupportedOperationException("Implement climbStairs");
    }
    public static void main(String[] args) {
        HouseRobberEasy s = new HouseRobberEasy();
        assert s.climbStairs(2) == 2;
        assert s.climbStairs(3) == 3;
        System.out.println("All test cases passed!");
    }

}

File: Blind75-Practice/easy/15-1d-dp/arrays/Python/house-robber-easy.py
"""
Easy Version: Climbing Stairs
Difficulty: Easy
Pattern: Dynamic Programming

You are climbing a staircase. It takes n steps to reach the top.
Each time you can either climb 1 or 2 steps. In how many distinct ways can you climb to the top?

Example 1:
Input: n = 2
Output: 2
Explanation: There are two ways to climb to the top.
1. 1 step + 1 step
2. 2 steps

Example 2:
Input: n = 3
Output: 3
Explanation: There are three ways to climb to the top.
1. 1 step + 1 step + 1 step
2. 1 step + 2 steps
3. 2 steps + 1 step

Time: O(n), Space: O(1)
"""

from typing import List


def climb_stairs(n: int) -> int:
    """
    Dynamic programming with space optimization
    Time: O(n), Space: O(1)
    """
    # TODO: Implement DP solution
    pass


# Test cases
if __name__ == "__main__":
    # Test 1
    n1 = 2
    print(f"Climb Stairs: {climb_stairs(n1)}")  # 2

    # Test 2
    n2 = 3
    print(f"Climb Stairs: {climb_stairs(n2)}")  # 3
